javascript/ql/src/semmle/javascript/Promises.qll

```diff
@@ -32,6 +32,39 @@ module Bluebird {
   }
 }
 
+/**
+ * Provides classes for working with various Deferred implementations
+ */
+module Deferred {
+  private DataFlow::SourceNode deferred() {
+    exists(VarAccess var, DataFlow::NewNode instantiation |
+      var.getName() = "Deferred" and
```

Comment:
This has to be a more semantic definition. We have heuristics in `.ql` files, and very, very, rarely in `.qll` files. This separation principle is in place to prevent a slow, catastrophic erosion of precision caused by interplay between multiple heuristics in the library files. I am glad to see the sanity check, but I do not think it is enough.

Some thoughts: 

Can you perhaps use `globalVarRef("Deferred")` instead? We use `globalVarRef` in a few places if it is common that the object of interest is introduced globally instead of through an imported library.

I note that something like [class Deferred](https://lgtm.com/projects/g/ampproject/amphtml/snapshot/6346a7ad0dd78ab949a4b9c81a87e684cd999b66/files/src/utils/promise.js#L38) is used in a few places in your query console hits. I am inclined to accept small class definitions (perhaps also pseudo class definitions with object literals and/or prototype hackery) with both `resolve` and `reject` fields as sources for deferred objects, but it depends on the results. These clases should also have at least one instance with your `exists(instantiation.getAMemberCall("resolve"))` heuristic, and perhaps also at least one instance with `exists(instantiation.getAMemberCall("reject"))`. It may be necessary to use type *tracking* to identify such classes properly.

Reply:
> Can you perhaps use globalVarRef("Deferred") instead? 

Not really. It happens often enough that `Deferred` is not a global variable (including in the FP that motivated this PR).   
I might be able to use `globalVarRef("Deferred")` together with `any(ParameterNode p | p.getName() = "Deferred")` to capture most uses.  
I'll look into creating something more precise.

> I note that something like class Deferred is used in a few places in your query console hits.  

But I don't think its used often enough to rely solely on that pattern.

Reply:
So I identified the patterns used within the libraries and changed the model accordingly: https://github.com/Semmle/ql/pull/2102/commits/4ec825b5b6f3dbb1ba8c5e20bf70963772c9bda0 

I don't think its much better than before, but it more precisely describes the patterns observed.

Reply:
> I note that something like [class Deferred](https://lgtm.com/projects/g/ampproject/amphtml/snapshot/6346a7ad0dd78ab949a4b9c81a87e684cd999b66/files/src/utils/promise.js#L38) is used in a few places in your query console hits. I am inclined to accept small class definitions (perhaps also pseudo class definitions with object literals and/or prototype hackery) with both `resolve` and `reject` fields as sources for deferred objects, but it depends on the results. These clases should also have at least one instance with your `exists(instantiation.getAMemberCall("resolve"))` heuristic, and perhaps also at least one instance with `exists(instantiation.getAMemberCall("reject"))`. It may be necessary to use type _tracking_ to identify such classes properly.

Something like this: https://github.com/Semmle/ql/commit/31009d979d323eb496530ae99c9f719ebed67272? 

I still don't check if the "class"-definition itself has `resolve`/`reject` methods. Mostly because I can't do that check when `Deferred` is a parameter.